### PR TITLE
ci: enable smoke tests for supported OS

### PR DESCRIPTION
**Pull request by v1v**
## Motivation/summary

Run the implemented [smoke tests for supported operating systems](https://github.com/elastic/apm-server/pull/9440) in the CI when:

- `daily`
- `manually`

Run the smoke tests in a dedicated CI pipeline to target the latest `SNAPSHOT` versions for:
- `main`
- latest release branch
- latest upcoming release branch, aka BC

It also adds tags/labels to the Cloud resources so it can help to search for long running resources and what/when they were created. 

## Fixes

```
[2022-11-21T13:55:32.707Z] Program: /bin/sh
[2022-11-21T13:55:32.707Z] Error Message: ./latest_apm_server.sh: 3: set: Illegal option -o pipefail
```

and

```
[2022-11-21T12:37:50.227Z] parse error: Invalid numeric literal at line 1, column 5
[2022-11-21T12:37:50.227Z] + VERSIONS=
```

## How to test these changes

Within the CI

## Related issues

https://github.com/elastic/apm-server/pull/9440

## Errors

```
[2022-11-21T14:24:04.965Z] Error: error configuring Terraform AWS Provider: error validating provider credentials: error calling sts:GetCallerIdentity: operation error STS: GetCallerIdentity, failed to resolve service endpoint, an AWS region is required, but was not found
```

## Follow ups

- Run on branches if new changes only.

## Tests

~~See https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fsmoke-tests-os-mbp/detail/PR-9624/9/pipeline failed with~~ (It's now solved)

```
[2022-11-21T15:17:39.591Z] Error: failed tracking create progress: 2 errors occurred:
[2022-11-21T15:17:39.591Z] 	* found deployment plan errors: deployment [0742e618660b105d447af61a4e8e5f09] - [integrations_server][2e496628f69d406b87bba839e3ab6795]: caught error: "Plan change failed: Some instances were not running"
[2022-11-21T15:17:39.591Z] 	* set "request_id" to "g25bj477sllnn5rhhy96rujpmwcd2d4utf8bm206kgvheggbn6eiav46r9xzghg8" to recreate the deployment resources
[2022-11-21T15:17:39.591Z]   with module.ec_deployment.ec_deployment.deployment,
[2022-11-21T15:17:39.591Z]   on ../../infra/terraform/modules/ec_deployment/deployment.tf line 13, in resource "ec_deployment" "deployment":
[2022-11-21T15:17:39.591Z]   13: resource "ec_deployment" "deployment" {
```

https://apm-ci.elastic.co/job/apm-server/job/smoke-tests-os-mbp/view/change-requests/job/PR-9624/16/ failed with:

```
[2023-01-09T11:25:26.775Z] -> Asserting logs-apm.error-* contains expected documents documents...
[2023-01-09T11:25:26.775Z] Didn't find 1 indexed documents error.id=9876543210abcdeffedcba0123456789, total hits 0
[2023-01-09T11:25:26.775Z] {"took":0,"timed_out":false,"_shards":{"total":0,"successful":0,"skipped":0,"failed":0},"hits":{"total":{"value":0,"relation":"eq"},"max_score":0.0,"hits":[]}}
[2023-01-09T11:25:26.776Z] -> Smoke tests FAILED!!
```

### CODE REVIEW

File: testing/smoke/supported-os/main.tf
Comment (v1v) [inline testing/smoke/supported-os/main.tf:201]: > We're using the CFT region on ESS (gcp-us-west2) so we might want to use something in the same region in AWS to reduce cost/bandwith

File: .ci/schedule-daily.groovy
Comment (v1v) [inline .ci/schedule-daily.groovy:25]: Enable the smoke tests for supported Operating Systems on a daily basis for the current release branches and main.

File: .ci/smoke-tests-os.groovy
Comment (v1v) [inline .ci/smoke-tests-os.groovy:12]:
```suggestion
    SLACK_CHANNEL = "#apm-server"
```

File: .ci/smoke-tests-os.groovy
Comment (v1v) [inline .ci/smoke-tests-os.groovy:37]: Gather the curren APM-Server version so it can trigger the smoke tests against the correct version later on.

File: .ci/smoke-tests-os.groovy
Comment (v1v) [inline .ci/smoke-tests-os.groovy:46]: Use snapshots rather than releases, `make get-version` returns major.minor.patch, so let's add `-SNAPSHOT` so the `testing/smoke/supported-os/test.sh` can use that version.

this is handy to support main and the release branches.

**Review by cachedout — APPROVED** [review]

**Comment by marclop** [comment]
I copied some of the changes int his PR to fix the smoke tests since they were failing on main. I think this needs to be accounted here.

**Comment by simitt** [comment]
@v1v is this PR currently blocked by anything from the APM Server team?

**Comment by v1v** [comment]
> @v1v is this PR currently blocked by anything from the APM Server team?

I wanted some 👀 before merging it. I've just rebased this PR and this [build](https://apm-ci.elastic.co/job/apm-server/job/smoke-tests-os-mbp/view/change-requests/job/PR-9624/16/) will validate the changes, if it gets success, then I'll ask for some reviews before merging it

**Comment by v1v** [comment]
The smoke tests are running but the assertions are failing 

```
[2023-01-09T11:25:26.775Z] -> Asserting logs-apm.error-* contains expected documents documents...
[2023-01-09T11:25:26.775Z] Didn't find 1 indexed documents error.id=9876543210abcdeffedcba0123456789, total hits 0
[2023-01-09T11:25:26.775Z] {"took":0,"timed_out":false,"_shards":{"total":0,"successful":0,"skipped":0,"failed":0},"hits":{"total":{"value":0,"relation":"eq"},"max_score":0.0,"hits":[]}}
[2023-01-09T11:25:26.776Z] -> Smoke tests FAILED!!
```

I assume this is not related to this particular changes as it runs only in the CI, but something else

**Comment by kruskall** [comment]
> The smoke tests are running but the assertions are failing
> 
> ```
> [2023-01-09T11:25:26.775Z] -> Asserting logs-apm.error-* contains expected documents documents...
> [2023-01-09T11:25:26.775Z] Didn't find 1 indexed documents error.id=9876543210abcdeffedcba0123456789, total hits 0
> [2023-01-09T11:25:26.775Z] {"took":0,"timed_out":false,"_shards":{"total":0,"successful":0,"skipped":0,"failed":0},"hits":{"total":{"value":0,"relation":"eq"},"max_score":0.0,"hits":[]}}
> [2023-01-09T11:25:26.776Z] -> Smoke tests FAILED!!
> ```
> 
> I assume this is not related to this particular changes as it runs only in the CI, but something else. @kruskall , do you happen to know what's the reason for the above error when running the smoke tests?

Sorry for the late response, I'm looking into this but it seems there is something strange going on. I tried to query the error index in ES but there are no documents. I'll keep investigating.

**Comment by v1v** [comment]
@kruskall, did you find anything that explained those errors? I reran it again and it failed with a different error

see [this](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fsmoke-tests-os-mbp/detail/PR-9624/17/pipeline#step-177-log-46)

```


[2023-01-18T11:36:13.376Z] -> Running basic supported OS smoke test for version 8.7.0-SNAPSHOT
[2023-01-18T11:36:13.376Z] -> Creating / Upgrading deployment to version 
[2023-01-18T11:41:01.229Z] -> Removing all data from ES...
[2023-01-18T11:41:01.531Z] {"acknowledged":true}-> APM Server ready!
[2023-01-18T11:41:01.531Z] -> Sending events to APM Server...
[2023-01-18T11:41:11.771Z] -> Asserting component templates and ILM policies...
[2023-01-18T11:41:18.094Z] -> Asserting ingest pipelines...
[2023-01-18T11:41:18.496Z] -> Asserting logs-apm.error-* contains expected documents documents...
[2023-01-18T11:41:18.496Z] -> Asserted 1 error.id=9876543210abcdeffedcba0123456789 exists
[2023-01-18T11:41:18.697Z] -> Asserting traces-apm-* contains expected documents documents...
[2023-01-18T11:41:18.697Z] -> Asserted 1 span.id=1234567890aaaade exists
[2023-01-18T11:41:18.899Z] -> Asserting traces-apm-* contains expected documents documents...
[2023-01-18T11:41:18.899Z] -> Asserted 1 transaction.id=4340a8e0df1906ecbfa9 exists
[2023-01-18T11:41:19.101Z] -> Asserting metrics-apm.internal-* contains expected documents documents...
[2023-01-18T11:41:19.101Z] -> Asserted 1 transaction.type=request exists
[2023-01-18T11:41:19.101Z] -> Creating / Upgrading deployment to version 
[2023-01-18T11:43:00.100Z] -> Removing all data from ES...
[2023-01-18T11:45:10.189Z] {"acknowledged":true}-> APM Server isn't ready to receive events, retrying (1/10)...
[2023-01-18T11:47:21.268Z] -> APM Server isn't ready to receive events, retrying (2/10)...
...
[2023-01-18T12:05:28.779Z] -> APM Server isn't ready to receive events, maximum retries exceeded
[2023-01-18T12:05:28.780Z] -> Smoke tests FAILED!!
```


I'm trying again with [this build](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fsmoke-tests-os-mbp/detail/PR-9624/18/pipeline)

**Comment by kruskall** [comment]
Discusses on slack, here are my findings:

Index templates have been renamed in https://github.com/elastic/apm-server/commit/4bc2e1a7aeb7ce896b8f552a3f2a279b9436913d but it seems the old index templates are still being created on ES. This is failing the 'apm integration installed' check which is breaking APM Server standalone.

Relevant output from the APM Server logs from the smoke testing:

```
Jan 29 04:15:45 ip-172-31-5-130.eu-central-1.compute.internal apm-server[4647]: {"log.level":"error","@timestamp":"2023-01-29T04:15:45.156Z","log.logger":"beater","log.origin":{"file.name":"beater/waitready.go","file.line":62},"message":"precondition 'apm integration installed' failed: error querying Elasticsearch for integration index templates: unexpected HTTP status: 404 Not Found ({\"error\":{\"root_cause\":[{\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching [metrics-apm.service_transaction.60m] not found\"}],\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching [metrics-apm.service_transaction.60m] not found\"},\"status\":404}): to remediate, please install the apm integration: https://ela.st/apm-integration-quickstart","service.name":"apm-server","ecs.version":"1.6.0"}
Jan 29 04:15:50 ip-172-31-5-130.eu-central-1.compute.internal apm-server[4647]: {"log.level":"error","@timestamp":"2023-01-29T04:15:50.132Z","log.logger":"beater","log.origin":{"file.name":"beater/waitready.go","file.line":62},"message":"precondition 'apm integration installed' failed: error querying Elasticsearch for integration index templates: unexpected HTTP status: 404 Not Found ({\"error\":{\"root_cause\":[{\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching [metrics-apm.service_transaction.10m] not found\"}],\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching [metrics-apm.service_transaction.10m] not found\"},\"status\":404}): to remediate, please install the apm integration: https://ela.st/apm-integration-quickstart","service.name":"apm-server","ecs.version":"1.6.0"}
Jan 29 04:15:55 ip-172-31-5-130.eu-central-1.compute.internal apm-server[4647]: {"log.level":"error","@timestamp":"2023-01-29T04:15:55.130Z","log.logger":"beater","log.origin":{"file.name":"beater/waitready.go","file.line":62},"message":"precondition 'apm integration installed' failed: error querying Elasticsearch for integration index templates: unexpected HTTP status: 404 Not Found ({\"error\":{\"root_cause\":[{\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching [metrics-apm.service_transaction.1m] not found\"}],\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching [metrics-apm.service_transaction.1m] not found\"},\"status\":404}): to remediate, please install the apm integration: https://ela.st/apm-integration-quickstart","service.name":"apm-server","ecs.version":"1.6.0"}
```

Output from ES Get index template API (curl -s -u user:pass https://elasticsearch/_index_template | jq ".index_templates[].name" | sort):

```
"apm-source-map"
".deprecation-indexing-template"
".fleet-file-data"
".fleet-files"
"ilm-history"
".kibana-event-log-8.7.0-snapshot-template"
"logs"
"logs-apm.app"
"logs-apm.error"
"metrics"
"metrics-apm.app"
"metrics-apm.internal"
"metrics-apm.service.10m"
"metrics-apm.service.1m"
"metrics-apm.service.60m"
"metrics-apm.service_destination.10m"
"metrics-apm.service_destination.1m"
"metrics-apm.service_destination.60m"
"metrics-apm.service_summary.10m"
"metrics-apm.service_summary.1m"
"metrics-apm.service_summary.60m"
"metrics-apm.transaction.10m"
"metrics-apm.transaction.1m"
"metrics-apm.transaction.60m"
".ml-anomalies-"
".ml-notifications-000002"
".ml-state"
".ml-stats"
".monitoring-beats-mb"
".monitoring-ent-search-mb"
".monitoring-es-mb"
".monitoring-kibana-mb"
".monitoring-logstash-mb"
".slm-history"
"synthetics"
"synthetics-browser"
"synthetics-browser.network"
"synthetics-browser.screenshot"
"synthetics-http"
"synthetics-icmp"
"synthetics-tcp"
"traces-apm"
"traces-apm.rum"
"traces-apm.sampled"
".watch-history-16"
```

From the logs of the ESS deployment:

```
[instance-0000000000] adding index template [metrics-apm.service.60m] for index patterns [metrics-apm.service.60m-*]
[instance-0000000000] adding index template [metrics-apm.service.1m] for index patterns [metrics-apm.service.1m-*]
[instance-0000000000] adding index template [metrics-apm.service.10m] for index patterns [metrics-apm.service.10m-*]
```

We're still investigating this

**Comment by kruskall** [comment]
@v1v Sorry for the long wait, this was a fun one. The issues should be fixed once https://github.com/elastic/apm-server/pull/10217 is merged

**Comment by kruskall** [comment]
> @v1v Sorry for the long wait, this was a fun one. The issues should be fixed once #10217 is merged

PR Merged. Smoke tests should work now

**Comment by simitt** [comment]
@v1v @kruskall can you drive this to resolution - the PR has been open for quite some time, and afaik all dependencies are resolved now.

**Comment by v1v** [comment]
I've just updated the PR and rerun the smoke tests:
- https://apm-ci.elastic.co/job/apm-server/job/smoke-tests-os-mbp/job/PR-9624/20/

**Review by kruskall — APPROVED** [review]

**Pull request merged**
